### ISSUE TITLE
spurious CI error for MSVC

### ISSUE THREAD

**Issue by ehuss** [rust-lang/rust#81890]
There have been a few instances of this error happening on CI:

>   = note: LINK : fatal error LNK1201: error writing to program database 'D:\a\rust\rust\build\i686-pc-windows-msvc\test\run-make-fulldeps\extern-flag-fun\extern-flag-fun\foo.pdb'; check for insufficient disk space, invalid path, or insufficient privilege

Or some variant of that.  See:

* https://github.com/rust-lang/rust/pull/81257#issuecomment-771231808
* https://github.com/rust-lang/rust/pull/81411#issuecomment-774824226
* https://github.com/rust-lang/rust/pull/80802#issuecomment-759450136
* https://github.com/rust-lang/rust/pull/80591#issuecomment-753535696
* https://github.com/rust-lang/rust/pull/80209#issuecomment-751316136
* https://github.com/rust-lang/rust/pull/80122#issuecomment-747612121
* https://github.com/rust-lang/rust/pull/80009#issuecomment-754213646
* https://github.com/rust-lang/rust/pull/80003#issuecomment-745227156

(and probably many more)

**Comment by kennytm** [comment]
Not just i686.

* https://github.com/rust-lang/rust/pull/90542#issuecomment-967741298
* https://github.com/rust-lang/rust/pull/90684#issuecomment-966792302

**Comment by hkratz** [comment]
The cargo test `lto::complicated` has been triggering this a lot lately. And cargo itself prints warnings, because the same .pdb file is written multiple times:
```
021-11-28T17:13:11.5135303Z The bin target `test` in package `test v0.0.0 (D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\tmp\cit\t1300\foo)` has the same output filename as the lib target `test` in package `test v0.0.0 (D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\tmp\cit\t1300\foo)`.
2021-11-28T17:13:11.5137366Z Colliding filename is: D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\tmp\cit\t1300\foo\target\release\deps\test.pdb
2021-11-28T17:13:11.5138476Z The targets should have unique names.
2021-11-28T17:13:11.5139250Z Consider changing their names to be unique or compiling them separately.
2021-11-28T17:13:11.5140516Z This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
2021-11-28T17:13:11.5141536Z warning: output filename collision.
2021-11-28T17:13:11.5143164Z The bin target `test` in package `test v0.0.0 (D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\tmp\cit\t1300\foo)` has the same output filename as the lib target `test` in package `test v0.0.0 (D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\tmp\cit\t1300\foo)`.
2021-11-28T17:13:11.5145380Z Colliding filename is: D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\tmp\cit\t1300\foo\target\release\test.pdb
2021-11-28T17:13:11.5146482Z The targets should have unique names.
2021-11-28T17:13:11.5147438Z Consider changing their names to be unique or compiling them separately.
2021-11-28T17:13:11.5148495Z This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
[...]
021-11-28T17:13:11.5246422Z   = note: LINK : fatal error LNK1201: error writing to program database 'D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\tmp\cit\t1300\foo\target\release\deps\test.pdb'; check for insufficient disk space, invalid path, or insufficient privilege
2021-11-28T17:13:11.5247882Z           
2021-11-28T17:13:11.5248143Z 
2021-11-28T17:13:11.5248698Z error: could not compile `test` due to previous error
```
So I guess _something_ locks the file after it is written so it can't be overwritten shortly thereafter. Can the test be updated not to do this?

**Comment by ehuss** [comment]
@hkratz I'm looking into fixing that test (and doing some other collision stuff).

I'm also trying to dig into `LNK1201`. Most of the errors seem to be on tests that build the same crate twice in a row. Unfortunately everything I've tried to reproduce it haven't succeeded.

Just on the off chance we can find someone who wants to help:

@rustbot ping windows

This issue is causing significant trouble on CI, and I'm having a hard time making progress on it.  A huge help would be if someone can find some way to reproduce it.  I've tried running `rustc -g` in a loop, but I can't trigger the error.  A really trivial test that triggered this error in the past is https://github.com/rust-lang/rust/blob/master/src/test/run-make-fulldeps/override-aliased-flags/Makefile.

**Comment by hkratz** [comment]
> @hkratz I'm looking into fixing that test (and doing some other collision stuff).

@ehuss Great! I wrote a little script to grep through the CI logs and of the last 30 failed msvc runs which had LNK1201 in it *all* failed at `lto::complicated`. So fixing this test will solve this for us.

The script for reference (excuse my Python, not a Python native):
```python
import github
import requests
from github import Github


def main():
    token = "<replace with token>"
    g = Github(token)
    # github.enable_console_debug_logging()
    repo_name = "rust-lang-ci/rust"
    repo = g.get_repo(repo_name)
    workflow_runs = repo.get_workflow_runs(status="failure", branch="auto")
    for run in workflow_runs:
        jobs_url = run.jobs_url + "?per_page=100"
        response = requests.get(jobs_url, headers={"Authorization": f"token {token}"})
        response.raise_for_status()
        for job in response.json()["jobs"]:
            if job["conclusion"] == "failure" and "msvc" in job["name"]:
                log_url = f"https://api.github.com/repos/{repo_name}/actions/jobs/{ job['id']}/logs"
                log_response = requests.get(
                    log_url, headers={"Authorization": f"token {token}"}
                )
                log_response.raise_for_status()
                log = log_response.text
                if "LNK1201" in log:
                    print(log_url)
                    print("\n\n")
                    print(log)
```

**Comment by wesleywiser** [comment]
> So I guess _something_ locks the file after it is written so it can't be overwritten shortly thereafter.

I talked with an engineer internally and they said that LNK1201 most often occurs because some program has a lock on the file. I don't know what the configuration of the GH runners looks like but perhaps Windows Defender (or similar AV/AM software) is scanning the file right after the linker finishes writing it the first time?

Regardless, rust-lang/cargo#10137 looks like the correct approach to resolving this to me. Thanks @ehuss!

**Comment by agersant** [comment]
I have a codebase where the [CI workflow on Github Actions](https://github.com/agersant/tiger/actions/workflows/ci.yml) has started exhibiting this error on every run (5/5 so far, starting with run #307). This is failing very early in the build so I do not think the Github Actions runner is running out of disk space. Is there any useful information that could be extracted from this somewhat reproducible setup?

**Comment by ematipico** [comment]
I tried to upgrade the toolchain to rust `1.72.0` and I hit this issue: https://github.com/biomejs/biome/actions/runs/5975779047/job/16212389309?pr=66

Interestingly, this issue occurs only when running tests using `nextest`.

**Comment by nandin-borjigin** [comment]
> Regardless, [rust-lang/cargo#10137](https://github.com/rust-lang/cargo/pull/10137) looks like the correct approach to resolving this to me. Thanks @ehuss!

@wesleywiser, did you mean [rust-lang/cargo#10137] is a fix only to the CI error mentioned in this issue, or is it supposed to fix the transient LNK1201 errors (due to something, most likely MsDefender, locking the artifacts) for all cases? 

I assume you meant the former, right? Do we have an issue tracking the LNK1201 that happens even though there is not colliding artifacts?

**Comment by ehuss** [comment]
I'm going to close as I do not think we have been hitting LNK1201 errors in a while. I don't know what happened with the `extern-flag-fun` test, but hopefully this is a thing of the past.
